packages/components/src/circular-option-picker/circular-option-picker-option.tsx

```diff
@@ -46,18 +45,18 @@ function UnforwardedOptionAsOption(
 		id: string;
 		className?: string;
 		isSelected?: boolean;
-		compositeStore: NonNullable<
-			React.ComponentProps< typeof Composite >[ 'store' ]
-		>;
 	},
 	forwardedRef: ForwardedRef< any >
 ) {
-	const { id, isSelected, compositeStore, ...additionalProps } = props;
-	const activeId = useStoreState( compositeStore, 'activeId' );
+	const { id, isSelected, ...additionalProps } = props;
 
-	if ( isSelected && ! activeId ) {
-		compositeStore.setActiveId( id );
-	}
+	const { setActiveId, activeId } = useContext( CircularOptionPickerContext );
+
+	useEffect( () => {
+		if ( isSelected && ! activeId ) {
+			window.setTimeout( () => setActiveId?.( id ), 0 );
```

Comment:
Without this `setTimeout`, the `setActiveId` call would be immediately overridden by Composite's internal logic that picks the first item as the active item (if no initial active item is passed). Not sure if there's a better way.

Reply:
`window.queueMicrotask` should do? It's not necessarily better, but maybe a bit more intentional.

Reply:
I can try that in https://github.com/WordPress/gutenberg/pull/64833 if we think we should move on with the approach

Reply:
Do we need a test for this?

I fiddled a bit and found that having access to `setItems` will avoid the need for a timeout. Despite the additional lines of code, I _think_ this is preferable in terms of what we can recommend to extenders. "Use a timeout/rAF/queueMicrotask in this specific scenario" is way too random and imperative for us to write in docs 😅 Also, I think we will encounter use cases where we need access to `items` anyway, in lieu of store methods like `first()` and `last()`.

```diff
diff --git a/packages/components/src/circular-option-picker/circular-option-picker-option.tsx b/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
index 1136f659b1..fb251f8c53 100644
--- a/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
+++ b/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
@@ -54,7 +54,7 @@ function UnforwardedOptionAsOption(
 
 	useEffect( () => {
 		if ( isSelected && ! activeId ) {
-			window.setTimeout( () => setActiveId?.( id ), 0 );
+			setActiveId?.( id );
 		}
 	}, [ isSelected, setActiveId, activeId, id ] );
 
diff --git a/packages/components/src/circular-option-picker/circular-option-picker.tsx b/packages/components/src/circular-option-picker/circular-option-picker.tsx
index adf2b386e5..55f17903a7 100644
--- a/packages/components/src/circular-option-picker/circular-option-picker.tsx
+++ b/packages/components/src/circular-option-picker/circular-option-picker.tsx
@@ -8,7 +8,7 @@ import clsx from 'clsx';
  */
 import { useInstanceId } from '@wordpress/compose';
 import { isRTL } from '@wordpress/i18n';
-import { useMemo, useState } from '@wordpress/element';
+import { useEffect, useMemo, useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -89,6 +89,7 @@ function ListboxCircularOptionPicker(
 	const [ activeId, setActiveId ] = useState< string | null | undefined >(
 		undefined
 	);
+	const [ items, setItems ] = useState();
 
 	const contextValue = useMemo(
 		() => ( {
@@ -99,6 +100,12 @@ function ListboxCircularOptionPicker(
 		[ baseId, activeId, setActiveId ]
 	);
 
+	useEffect( () => {
+		if ( items && ! activeId ) {
+			setActiveId( items[ 0 ].id );
+		}
+	}, [ activeId, items ] );
+
 	return (
 		<div className={ className }>
 			<CircularOptionPickerContext.Provider value={ contextValue }>
@@ -110,6 +117,8 @@ function ListboxCircularOptionPicker(
 					role="listbox"
 					activeId={ activeId }
 					setActiveId={ setActiveId }
+					defaultActiveId={ null }
+					setItems={ setItems }
 				>
 					{ options }
 				</Composite>
diff --git a/packages/components/src/composite/index.tsx b/packages/components/src/composite/index.tsx
index e8e3a5811a..ce8c0575e3 100644
--- a/packages/components/src/composite/index.tsx
+++ b/packages/components/src/composite/index.tsx
@@ -56,6 +56,7 @@ export const Composite = Object.assign(
 			activeId,
 			defaultActiveId,
 			setActiveId,
+			setItems,
 			focusLoop = false,
 			focusWrap = false,
 			focusShift = false,
@@ -81,6 +82,7 @@ export const Composite = Object.assign(
 			focusShift,
 			virtualFocus,
 			orientation,
+			setItems,
 			rtl,
 		} );


```
(☝️ Rough sketch code, isn't typed and doesn't filter disabled items)

Tangent: We have this kind of `activeId` syncing logic in Tabs, too, which involves some `rAF` shenanigans. Maybe there is some reworking that can be done to avoid that as well 🤔 But I'm more forgiving of that because that's internal code and it's not like we'd need to tell extenders to do it.

Reply:
Good point. Maybe I'll ping Diego in https://github.com/WordPress/gutenberg/pull/64833 and ask him what is, in his  opinion, a better approach to using timeouts/rAFs/queueMicrotasks